sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsPartitions.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.connector.catalog;
+
+import java.util.Map;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * Catalog methods for working with Partitions.
+ */
+@Experimental
+public interface SupportsPartitions extends TableCatalog {
+
+    /**
+     * Create partitions in an existing table, assuming it exists.
+     *
+     * @param ident a table identifier
+     * @param partitions transforms to use for partitioning data in the table
+     * @param ignoreIfExists
+     */
+    void createPartitions(
+            Identifier ident,
+            TablePartition[] partitions,
+            Boolean ignoreIfExists);
+
+    /**
+     * Drop partitions from a table, assuming they exist.
+     *
+     * @param ident a table identifier
+     * @param partitions a list of string map for existing partitions
+     * @param ignoreIfNotExists
+     */
+    void dropPartitions(
+            Identifier ident,
+            Map<String, String>[] partitions,
```

Comment:
There are a few cases here where partitions are referred to as Map<String, String> and a few times where they use the TablePartition class. I think it would probably make more sense if they were all TablePartition (the class) unless there is a significant reason for them not to be.

Reply:
TablePartition Contains the partition metadata, it's too heavy for this. As for Transform, it maybe a good choice if it can pass partition value.

Reply:
Then is this partitionSpec and not "partition"?

Reply:
I think we need to decide how to pass the data that defines a partition.

There have been a _lot_ of problems over the years working with Hive partitions because values are coerced to and from String. Often, people get the conversions slightly wrong. I think a better approach is to use a row of values to pass partition data between Spark and a source. We already pass typed rows in the read and write APIs, so it would be reasonable to do so here as well.

One benefit of using a typed row to represent partition data is that we can directly use a `listPartitions` call for metadata queries.

This would also align more closely with how Spark handles partitions internally. From PartitioningUtils:

```
/**
 * Holds a directory in a partitioned collection of files as well as the partition values
 * in the form of a Row.  Before scanning, the files at `path` need to be enumerated.
 */
case class PartitionPath(values: InternalRow, path: Path)

case class PartitionSpec(
    partitionColumns: StructType,
    partitions: Seq[PartitionPath])
```

A table that implements `SupportsPartitions` could return a `partitionType(): StructType` that describes the partition rows it accepts and produces.

Reply:
One more thing: using a row to pass a tuple would make it possible to also get the partition that an input split (e.g., file) belongs to. That would be useful for storage-partitioned joins.

Reply:
Good Point! The partition identifiers can be replaced as `PartitionSpec`, it is a better class for this.

Reply:
I was actually thinking that partitions would be identified more like `PartitionPath`. In this API, I'm not sure if the `Path` part of `PartitionPath` is needed, since sources may not need to expose it to Spark. (In Iceberg, for example, there is no partition path.)

I think just using an `InternalRow` to identify a partition is a good idea.

Reply:
Further think about this. `InternalRow` only contains the data of partition identifier, not partition columns. That means user must make partition data always in order. Is that reasonable to user? @rdblue

Reply:
I think it is reasonable. While this is a public API, the user here is a connector developer and they are expected to be able to produce `InternalRow` in other places. I think this is actually a good thing because we don't need to pass the names every time.